### PR TITLE
Update MSAL to proposed OM - HTTP Component + OAuth2 Client

### PR DESCRIPTION
**Pull request by kpanwar**
Update MSAL to reflect proposed OM - HTTP Component and OAuth2 Client. Use HttpRequest and HttpResponse model where request object is responsible for retries and is self contained with all the data to execute a HTTP request.
Create OAuth2 client that is responsible for parsing JSON and creating token response and instance discovery response objects. #108

### CODE REVIEW

File: src/MSAL.PCL/Internal/Http/MsalHttpRequest.cs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+using System.Globalization;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client.Internal.Http
+{
+    internal class MsalHttpRequest
```
Comment (addev-ashish) [inline src/MSAL.PCL/Internal/Http/MsalHttpRequest.cs]: Remove MSAL from the name. We know.

File: src/MSAL.PCL/Internal/Http/MsalHttpRequest.cs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            Dictionary<string, string> bodyParameters, HttpMethod method,
+            CallState callstate, bool retry = true)
+        {
+            bool isRetryable = false;
+            MsalHttpResponse response = null;
+            try
+            {
+                response = await Execute(endpoint, headers, bodyParameters, method);
+
+                PlatformPlugin.Logger.Error(callstate,
```
Comment (addev-ashish) [inline src/MSAL.PCL/Internal/Http/MsalHttpRequest.cs]: this should be in the else block.

File: src/MSAL.PCL/Internal/Http/HttpMessageHandlerFactory.cs [hunk 3]
```diff
@@ -25,23 +25,31 @@
...
-namespace Microsoft.Identity.Client.Internal
+namespace Microsoft.Identity.Client.Internal.Http
 {
-
     internal static class HttpMessageHandlerFactory
     {
-        internal static HttpMessageHandler GetMessageHandler(bool useDefaultCredentials)
+        private static readonly Queue<HttpMessageHandler> MockHttpMessageHandlerQueue = new Queue<HttpMessageHandler>();
+
+        internal static HttpMessageHandler GetMessageHandler(bool forMock)
```
Comment (addev-ashish) [inline src/MSAL.PCL/Internal/Http/HttpMessageHandlerFactory.cs]: consider renaming.

File: src/MSAL.PCL/Internal/OAuth2/OAuth2Client.cs [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+                _headers.Add(OAuth2Header.RequestCorrelationIdInResponse, "true");
+            }
+
+            MsalHttpResponse response =
+                await
+                    MsalHttpRequest.SendGet(CreateFullEndpointUri(endPoint), this._headers, callState);
+            return CreateResponse<InstanceDiscoveryResponse>(response, callState, addCorrelationId);
+        }
+
+        public async Task<TokenResponse> GetToken(Uri endPoint, CallState callState)
```
Comment (addev-ashish) [inline src/MSAL.PCL/Internal/OAuth2/OAuth2Client.cs]: collpase code for both functions.

File: src/MSAL.PCL/Internal/Http/MsalHttpRequest.cs (hunk 2 above)
Comment (kpanwar) [inline src/MSAL.PCL/Internal/Http/MsalHttpRequest.cs]: done

File: src/MSAL.PCL/Internal/Http/MsalHttpRequest.cs (hunk 1 above)
Comment (kpanwar) [inline src/MSAL.PCL/Internal/Http/MsalHttpRequest.cs]: done.

File: src/MSAL.PCL/Internal/Http/HttpMessageHandlerFactory.cs (hunk 3 above)
Comment (kpanwar) [inline src/MSAL.PCL/Internal/Http/HttpMessageHandlerFactory.cs]: leaving as is, as per the discussion

File: src/MSAL.PCL/Internal/OAuth2/OAuth2Client.cs (hunk 4 above)
Comment (kpanwar) [inline src/MSAL.PCL/Internal/OAuth2/OAuth2Client.cs]: done.

**Comment by weijjia** [comment]
:shipit:

**Pull request merged**